File: 2016/Rotate Function/Solution.cpp
#include <vector>
#include <stdio.h>
#include <limits.h>
using namespace std;
class Solution {
public:
    int maxRotateFunction(vector<int>& A) {
      int n = A.size();
      if (n == 0) {
        return 0;
      }
      int max = INT_MIN;
      for (int i = 0; i < n; i++) {
        int result = figure(A,n,i);
        if (max < result) {
          max = result;
        }
      }
      return max;
    }
    int figure(vector<int>& A, int length, int start) {
      int sum = 0;
      int i = 0;
      int index = start;
      while (i < length) {

        sum += A[index] * i;
        i = i + 1;
        index = index + 1;
        if (index >= length) {
          index = 0;
        }
      }
        return sum;
    }
};

File: 2016/Rotate Function/solution.py
class Solution(object):
    def maxRotateFunction(self, A):
        """
        :type A: List[int]
        :rtype: int
        """
        n = A.length();
        if n == 0:
            return 0;
        max = -1;
        for i in range(len(A)):
            result = figure(A,n,i);
            if max < result:
                max = result;
        return max;
    def figure(self, A, length, start):
        sum = 0;
        i = 0;
        index = start;
        while i < length:
            sum += A[index] * i;
            i = i + 1;
            index = index + 1;
            if index >= length:
                index = 0;

        return sum;

if __name__ == '__main__':
    Solution
